### DISCUSSION TITLE
Loading external assemblies at runtime

### DISCUSSION THREAD

**Discussion by Stridemann** [nxrighthere/UnrealCLR#65] [General]
Thanks for making this great project :) 

I got a problem with loading dll's from files in runtime (+using reflection on them later).
They seems loaded fine until they have a reference to UnrealEngine.Framework or UnrealEngine.Tests.

Loading using:
AssemblyLoadContext.Default.LoadFromAssemblyPath(fileName);
AssemblyLoadContext.Default.LoadFromStream(memoryStream);
Assembly.Load(memoryStream.ToArray());

I was also trying to use assembliesContext from AssembliesContextManager (in Runtime.cs)  instead AssemblyLoadContext.Default, but error is the same.

`LogUnrealCLR: Error: UnrealCLR::Module::Exception: System.Reflection.TargetInvocationException: Exception has been thrown by the target of an invocation.`
` ---> System.IO.FileNotFoundException: Could not load file or assembly 'UnrealEngine.Framework, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null'. ?? ??????? ????? ????????? ????.
File name: 'UnrealEngine.Framework, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null'
   at TestApi.TestCode.InvokeMe()
   --- End of inner exception stack trace ---
   at System.RuntimeMethodHandle.InvokeMethod(Object target, Object[] arguments, Signature sig, Boolean constructor, Boolean wrapExceptions)
   at System.Reflection.RuntimeMethodInfo.Invoke(Object obj, BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)
   at System.Reflection.MethodBase.Invoke(Object obj, Object[] parameters)
   at UnrealEngine.Tests.RuntimeCodeTest.test1()
   at UnrealEngine.Tests.RuntimeCodeTest.OnBeginPlay()
   at UnrealEngine.Tests.Main.OnWorldPostBegin()
   at UnrealEngine.Runtime.Core.ManagedCommand(Command command) in D:\HomeProjects\UE4\UnrealCLR\Source\Managed\Runtime\Runtime.cs:line 141`

I gonna help it to resolve assemblies using this:

AppDomain.CurrentDomain.AssemblyResolve += CurrentDomainOnAssemblyResolve;
   private Assembly? CurrentDomainOnAssemblyResolve(object? sender, ResolveEventArgs args)
        {
            if (args.Name.StartsWith("UnrealEngine.Tests"))
                return typeof(RuntimeCodeTest).Assembly;
            if (args.Name.StartsWith("UnrealEngine.Framework"))
                return typeof(Debug).Assembly;
            return null;
        }

but getting just [another exception](https://github.com/Potapy4/dotnet-coreclr/blob/master/src/vm/appdomain.cpp#L5953):
`LogUnrealCLR: Error: UnrealCLR::Module::Exception: System.Reflection.TargetInvocationException: Exception has been thrown by the target of an invocation.`
` ---> System.IO.FileLoadException: Could not load file or assembly 'UnrealEngine.Framework, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null'. Operation is not supported. (0x80131515)
File name: 'UnrealEngine.Framework, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null'
 ---> System.NotSupportedException: Resolving to a collectible assembly is not supported.
   at TestApi.TestCode.InvokeMe()
   --- End of inner exception stack trace ---
   at System.RuntimeMethodHandle.InvokeMethod(Object target, Object[] arguments, Signature sig, Boolean constructor, Boolean wrapExceptions)
   at System.Reflection.RuntimeMethodInfo.Invoke(Object obj, BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)
   at System.Reflection.MethodBase.Invoke(Object obj, Object[] parameters)
   at UnrealEngine.Tests.RuntimeCodeTest.test1()
   at UnrealEngine.Tests.RuntimeCodeTest.OnBeginPlay()
   at UnrealEngine.Tests.Main.OnWorldPostBegin()
   at UnrealEngine.Runtime.Core.ManagedCommand(Command command) in D:\HomeProjects\UE4\UnrealCLR\Source\Managed\Runtime\Runtime.cs:line 141`

Should I use some other way of loading.. like Runtime.cs do for referencedAssemblies ?
Thanks)

**Comment by Stridemann** [comment]
Also when I try to build UnrealEngine.Runtime to %MyProject%\Plugins\UnrealCLR\Managed I'm getting "Unloading of assemblies was failed!" without any code modification (I only changed output build path). Or how should I compile the runtime in a right way? Should I modify the runtime code and do Install again?

**Comment by nxrighthere** [comment]
What are you trying to achieve with that?

There should be only one referenced `UnrealEngine.Framework` assembly, and it can't be loaded dynamically from user code.

As for `UnrealEngine.Runtime`, it's not intended for user modifications once the plugin is installed.

**Comment by Stridemann** [comment]
I made a runtime code compilation (players can write, compile and execute code in the game). And it works fine. But I cannot use/reference to **UnrealEngine.Framework** or **UnrealEngine.Tests** (or some my custom assembly) in my runtime compiled code..
Actually only this problem stop me from moving forward in development..

```cs
private void TestRuntimeCompilation()
        {
            var compilation = CSharpCompilation.Create("c" + Guid.NewGuid())
                                               .WithOptions(new CSharpCompilationOptions(OutputKind.DynamicallyLinkedLibrary))
                                               .AddReferences(
                                                   MetadataReference.CreateFromFile(typeof(object).GetTypeInfo().Assembly.Location))
                                               .AddSyntaxTrees(CSharpSyntaxTree.ParseText(
                                                   "using System;\n" +
                                                   "using UnrealEngine.Framework;\n" +
                                                   "using System.Drawing;\n" +
                                                   "" +
                                                   "public class TestCode " +
                                                   "{" +
                                                   "    public static void InvokeMe()" +
                                                   "    {" +
                                                   "                Debug.Log(LogLevel.Display, \"=========Loaded!==========\");" +//remove this line to execute without exception
                                                   "    }" +
                                                   "}"));
            var dotnetCoreDirectory = System.Runtime.InteropServices.RuntimeEnvironment.GetRuntimeDirectory();
            compilation = compilation.AddReferences(MetadataReference.CreateFromFile(typeof(object).GetTypeInfo().Assembly.Location));
            compilation = compilation.AddReferences(MetadataReference.CreateFromFile(@"D:\HomeProjects\UE4\MyProject\Managed\Tests\UnrealEngine.Framework.dll")); //TODO: Fix path for ur side!
            compilation = compilation.AddReferences(MetadataReference.CreateFromFile(Path.Combine(dotnetCoreDirectory, "mscorlib.dll")));
            compilation = compilation.AddReferences(MetadataReference.CreateFromFile(Path.Combine(dotnetCoreDirectory, "netstandard.dll")));
            compilation = compilation.AddReferences(MetadataReference.CreateFromFile(Path.Combine(dotnetCoreDirectory, "System.Runtime.dll")));

            using (var memoryStream = new MemoryStream())
            {
                var emitResult = compilation.Emit(memoryStream);
                if (emitResult.Success)
                {
                    memoryStream.Seek(0, SeekOrigin.Begin);

                    var assembly = Assembly.Load(memoryStream.ToArray()); //context.LoadFromStream(memoryStream);
                    var type = assembly.GetType("TestCode");

                    Debug.Log(LogLevel.Error, "=========Executing...==========");
                    type.GetMethod("InvokeMe", BindingFlags.Public | BindingFlags.Static).Invoke(null, new object[] { });
                    Debug.Log(LogLevel.Error, "=========Executing done==========");
                }
                else
                {
                    var failures = emitResult.Diagnostics.Where(diagnostic => diagnostic.IsWarningAsError || diagnostic.Severity == DiagnosticSeverity.Error);
                    throw new Exception(string.Join(",", failures.Select(x => $"{x.Id}: {x.GetMessage()}")));
                }
            }
        }
```

```
LogUnrealManaged: Error: UnrealCLRFramework::Debug::Log: =========Executing...==========
LogUnrealCLR: Error: UnrealCLR::Module::Exception: System.Reflection.TargetInvocationException: Exception has been thrown by the target of an invocation.
 ---> System.IO.FileNotFoundException: Could not load file or assembly 'UnrealEngine.Framework, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null'. ?? ??????? ????? ????????? ????.
File name: 'UnrealEngine.Framework, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null'
   at TestCode.InvokeMe()
   --- End of inner exception stack trace ---
   at System.RuntimeMethodHandle.InvokeMethod(Object target, Object[] arguments, Signature sig, Boolean constructor, Boolean wrapExceptions)
   at System.Reflection.RuntimeMethodInfo.Invoke(Object obj, BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)
   at System.Reflection.MethodBase.Invoke(Object obj, Object[] parameters)
   at UnrealEngine.Tests.RuntimeCodeTest.TestRuntimeCompilation()
   at UnrealEngine.Tests.RuntimeCodeTest.OnBeginPlay()
   at UnrealEngine.Tests.Main.OnWorldPostBegin()
   at UnrealEngine.Runtime.Core.ManagedCommand(Command command) in D:\HomeProjects\UE4\UnrealCLR\Source\Managed\Runtime\Runtime.cs:line 141
```

**Comment by Stridemann** [comment]
If you remove this line it will execute without exception:
```"                Debug.Log(LogLevel.Display, \"=========Loaded!==========\");" +```

And to be able to run my code above you need to copy two dll's:
``Microsoft.CodeAnalysis.CSharp.dll``
``Microsoft.CodeAnalysis.dll``
to ``MyProject\Managed\Tests``

(you can get them [here](https://www.dropbox.com/s/drzuzg5lwmgy0s5/RequiredDllsForRuntimeCompilation.rar?dl=0) or copy from GAC)

**Comment by Stridemann** [comment]
We can check that required assembly is actually loaded in memory (obviously, in other case nothing works then). It just cannot resolve it in memory..

Log: Loaded assembly: System.Private.CoreLib, Version=5.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e
Log: Loaded assembly: UnrealEngine.Runtime, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null
Log: Loaded assembly: System.Runtime, Version=5.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a
Log: Loaded assembly: System.Runtime.InteropServices, Version=5.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a
Log: Loaded assembly: System.Runtime.Loader, Version=5.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a
Log: Loaded assembly: System.IO.FileSystem, Version=5.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a
Log: Loaded assembly: UnrealEngine.Plugins, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null
Log: Loaded assembly: System.Collections, Version=5.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a
Log: Loaded assembly: System.Linq, Version=5.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a
Log: Loaded assembly: System.Memory, Version=5.0.0.0, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51
Log: Loaded assembly: System.Threading, Version=5.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a
Log: Loaded assembly: UnrealEngine.Tests, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null
Log: Loaded assembly: System.IO.FileSystem.Watcher, Version=5.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a
Log: Loaded assembly: System.ComponentModel.Primitives, Version=5.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a
```Log: Loaded assembly: UnrealEngine.Framework, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null```
Log: Loaded assembly: System.Numerics.Vectors, Version=5.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a
Log: Loaded assembly: System.Reflection.Emit, Version=5.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a
Log: Loaded assembly: UnrealEngine.DynamicTypes, Version=0.0.0.0, Culture=neutral, PublicKeyToken=null
Log: Loaded assembly: System.Drawing.Primitives, Version=5.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a
Log: Loaded assembly: System.Runtime.Serialization.Formatters, Version=5.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a
Log: Loaded assembly: Microsoft.CodeAnalysis.CSharp, Version=3.7.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35
Log: Loaded assembly: Microsoft.CodeAnalysis, Version=3.7.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35
Log: Loaded assembly: System.Collections.Immutable, Version=5.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a
Log: Loaded assembly: System.Reflection.Metadata, Version=5.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a
Log: Loaded assembly: System.Threading.Tasks.Parallel, Version=5.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a
Log: Loaded assembly: System.Security.Cryptography.Algorithms, Version=5.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a
Log: Loaded assembly: System.Resources.ResourceManager, Version=5.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a
Log: Loaded assembly: System.Runtime.Extensions, Version=5.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a
Log: Loaded assembly: System.Security.Cryptography.Primitives, Version=5.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a
Log: Loaded assembly: System.Collections.Concurrent, Version=5.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a
Log: Loaded assembly: System.Text.Encoding.Extensions, Version=5.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a
Log: Loaded assembly: System.Globalization, Version=5.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a
Log: Loaded assembly: System.Diagnostics.Tracing, Version=5.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a
Log: Loaded assembly: System.Xml.XDocument, Version=5.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a
Log: Loaded assembly: System.Private.Xml.Linq, Version=5.0.0.0, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51
Log: Loaded assembly: System.Private.Xml, Version=5.0.0.0, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51
Log: Loaded assembly: System.Threading.Tasks, Version=5.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a
Log: Loaded assembly: System.Reflection.Primitives, Version=5.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a
```Log: Loaded assembly: ce79d8945-c9ea-4be7-9581-4a7bdf965761, Version=0.0.0.0, Culture=neutral, PublicKeyToken=null <<Runtime compiled dynamic assembly```


Log: =========Executing...==========
LogUnrealCLR: Error: UnrealCLR::Module::Exception: System.Reflection.TargetInvocationException: Exception has been thrown by the target of an invocation.
 ---> System.IO.FileNotFoundException: Could not load file or assembly '```UnrealEngine.Framework, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null'```. ?? ??????? ????? ????????? ????.
File name: 'UnrealEngine.Framework, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null'
   at TestCode.InvokeMe()
   --- End of inner exception stack trace ---
   at System.RuntimeMethodHandle.InvokeMethod(Object target, Object[] arguments, Signature sig, Boolean constructor, Boolean wrapExceptions)
   at System.Reflection.RuntimeMethodInfo.Invoke(Object obj, BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)
   at System.Reflection.MethodBase.Invoke(Object obj, Object[] parameters)
   at UnrealEngine.Tests.RuntimeCodeTest.TestRuntimeCompilation()
   at UnrealEngine.Tests.RuntimeCodeTest.OnBeginPlay()
   at UnrealEngine.Tests.Main.OnWorldPostBegin()
   at UnrealEngine.Runtime.Core.ManagedCommand(Command command) in D:\HomeProjects\UE4\UnrealCLR\Source\Managed\Runtime\Runtime.cs:line 141

**Comment by nxrighthere** [comment]
> If you remove this line it will execute without exception:
" Debug.Log(LogLevel.Display, \"=========Loaded!==========\");" +

Does this line persist in your external assembly which you are trying to load dynamically at runtime? Is that correct?

**Comment by Stridemann** [comment]
> > If you remove this line it will execute without exception:
> > " Debug.Log(LogLevel.Display, "=========Loaded!==========");" +
> 
> Does this line persist in your external assembly which you are trying to load dynamically at runtime? Is that correct?

Yes, it's in a [14 line of ``TestRuntimeCompilation`` function I posted above](https://github.com/nxrighthere/UnrealCLR/issues/65#issuecomment-737389605) which I'm compiling in-game

**Comment by Stridemann** [comment]
You can reference dll to ```UnrealEngine.Tests``` and call that code directly and it will work (obviously).
```zTestApi.TestCode.InvokeMe();```

But you cannot do anything with reflection/runtime loading:
```
var assembly = Assembly.Load(File.ReadAllBytes(@"D:\HomeProjects\UE4\SE\Managed\Tests\zTestApi.dll"));
assembly.GetType("zTestApi.TestCode").GetMethod("InvokeMe", BindingFlags.Public | BindingFlags.Static).Invoke(null, new object[] { });
```

dll code:

```cs
using UnrealEngine.Framework;
namespace zTestApi
{
    public class TestCode
    {
        public static void InvokeMe()
        {
            Debug.Log(LogLevel.Display, "=========Loaded!==========");
        }
    }
}

```

LogUnrealCLR: Error: UnrealCLR::Module::Exception: System.Reflection.TargetInvocationException: Exception has been thrown by the target of an invocation.
 ---> System.IO.FileNotFoundException: ```Could not load file or assembly 'UnrealEngine.Framework, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null'. ?? ??????? ????? ????????? ????.```
File name: 'UnrealEngine.Framework, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null'
   at zTestApi.TestCode.InvokeMe()
   --- End of inner exception stack trace ---
   at System.RuntimeMethodHandle.InvokeMethod(Object target, Object[] arguments, Signature sig, Boolean constructor, Boolean wrapExceptions)
   at System.Reflection.RuntimeMethodInfo.Invoke(Object obj, BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)
   at System.Reflection.MethodBase.Invoke(Object obj, Object[] parameters)
   at UnrealEngine.Tests.RuntimeCodeTest.test2()
   at UnrealEngine.Tests.RuntimeCodeTest.OnBeginPlay()
   at UnrealEngine.Tests.Main.OnWorldPostBegin()
   at UnrealEngine.Runtime.Core.ManagedCommand(Command command) in D:\HomeProjects\UE4\UnrealCLR\Source\Managed\Runtime\Runtime.cs:line 141

**Comment by nxrighthere** [comment]
As I mentioned previously, `UnrealEngine.Framework` can't be loaded dynamically from external assemblies. This assembly is designed specifically for the plugin with its internal functionality and loaded as a dependency only once after entering the play mode.

**Comment by nxrighthere** [comment]
In other words, you can't use framework API from assemblies that loaded dynamically from user code, that's why you get the same exception with `UnrealEngine.Tests` since `UnrealEngine.Framework` referenced there too.

**Comment by nxrighthere** [comment]
> I made a runtime code compilation (players can write, compile and execute code in the game).

You can try something like [Harmony](https://github.com/pardeike/Harmony) for that. It might work in this case.

**Comment by Stridemann** [comment]
I see, I got it. Thanks for a help)

> You can try something like [Harmony](https://github.com/pardeike/Harmony) for that. It might work in this case.

Thanks, I'll check again, but as far as I know it is used for patching the methods IL code..

Btw I was able to make everything work (also with `UnrealEngine.Framework`). I load my assembly in Runtime.cs in same way as Tests is loaded:
```
var myAsmName = AssemblyName.GetAssemblyName(@"D:\HomeProjects\UE4\SE\Managed\Tests\zTestApi.dll");
plugin.loader.LoadAssembly(myAsmName);
```
In worst case if I couldn't deal with Harmony I could make a wrapper/separate loader or assemblyLoadContext and just reload dlls like this in case they changed..


UPD. Runtime compilation in memory also works fine! (I'm using a PluginLoader.LoadContext to load the assemblies.. just for a test..)

**Comment by nxrighthere** [comment]
We can expose an interface for external plugins, if you have any idea how to do that properly, I'm open to any suggestions.

**Reply by Stridemann** [reply]
> We can expose an interface for external plugins if you have any idea how to do that properly, I'm open to any suggestions.

Constructor dependency injection. Like in Prism. We inject only if user need (by defining parameter in constructor). ~~But since we have static class it's impossible.~~
We can add optional function `public static void OnInitialize(IUnrealClrApi api)` to provide some interaction with core.
We can make a static helper class in Framework lib (this is better way).

Some of assemblies will be loaded once (let's say to default AssemblyLoadContext), another need to be reloaded (separate/own AssemblyLoadContext container)

```cs
Assembly LoadFromAssemblyPath(string assemblyPath)
Assembly LoadFromStream(Stream assembly)
Assembly LoadFromStream(Stream assembly, Stream? assemblySymbols)

public void Unload(Assembly assembly)//only trigger AssemblyLoadContext to unload, no GC and waiting
```

I could and want to take this task, but I need some time to play with and check how things is done in UnrealCLR..

**Reply by nxrighthere** [reply]
Sure, let me know how it's going for you.

**Reply by Stridemann** [reply]
> Sure, let me know how it's going for you.

UPD. Was trying so hard to stay on UE4, but moved to unity for a lot of reasons..(
Thanks for making this great project, it should live! It's a light in a darkness of C++ :)
